[shard] fix with_comms wrapper

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69493

When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
@with_comms
def test_case:
    ...
```
or
```
@with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Commit:
[shard] fix with_comms wrapper

When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
@with_comms
def test_case:
    ...
```
or
```
@with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/_sharded_tensor/__init__.py:
```diff
@@ -72,14 +72,18 @@ def assert_sharded_tensor_equal(self, st1, st2):
 
 # wrapper to initialize comms (processgroup + rpc)
-def with_comms(init_rpc=True, backend="nccl"):
+def with_comms(_func=None, *, init_rpc=True, backend="nccl"):
```
Wondering if we can just do what `dist_init` does: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/dist_utils.py#L20 to keep wrappers like these consistent.

Commit:
Update on "[shard] fix with_comms wrapper"


When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
with_comms
def test_case:
    ...
```
or
```
with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/_sharded_tensor/__init__.py:
```diff
@@ -72,14 +72,19 @@ def assert_sharded_tensor_equal(self, st1, st2):
-            self.destroy_comms(destroy_rpc=init_rpc)
-        return wrapper
-    return with_comms_decorator
+def with_comms(func=None, *, init_rpc=True, backend="nccl"):
```
nit: Do we need the '*' here?

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1834,64 +1840,66 @@ def test_init_from_local_shards_and_global_metadata(self):
-                shard_offsets=[0, 0],
-                shard_sizes=[5, 5],
-                placement="rank:0/cuda:1"
-            )
-            rank3_shard_metadata = ShardMetadata(
-                shard_offsets=[5, 0],
-                shard_sizes=[5, 5],
-                placement="rank:2/cuda:3"
-            )
+    #     if self.rank == 1 or self.rank == 3:
+    #         rank1_shard_metadata = ShardMetadata(
+    #             shard_offsets=[0, 0],
+    #             shard_sizes=[5, 5],
+    #             placement="rank:0/cuda:1"
```
Was this an unintended change?

Commit:
Update on "[shard] fix with_comms wrapper"


When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
with_comms
def test_case:
    ...
```
or
```
with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[shard] fix with_comms wrapper"


When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
with_comms
def test_case:
    ...
```
or
```
with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1840,58 +1846,34 @@ def test_init_from_local_shards_and_global_metadata(self):
                 self.assertEqual((5, 5), shard_metadata.shard_sizes)
-                self.assertEqual(f'rank:{rank * 2}/cuda:{rank * 2 + 1}', str(shard_metadata.placement))
-
-            # Validate remote shards.
-            remote_shards = sharded_tensor.remote_shards()
-            if self.rank == 1 or self.rank == 3:
-                self.assertEqual(1, len(remote_shards))
-            else:
-                self.assertEqual(2, len(remote_shards))
-
-            owners = {}
-            for rpc_rank, shards in remote_shards.items():
-                self.assertEqual(1, len(shards))
-
-                for remote_shard in shards:
-                    self.assertEqual(rpc_rank, remote_shard.owner().id)
-                    shard = remote_shard.to_here()
-                    self.assertEqual((5, 5), shard.tensor.size())
```
Don't we need to keep this validation for remote_shards?

wanchaol replied:
I think it's a bit tricky, the reason that I disabled this remote shards validation for this new_group test is that, we are essentially having two pg (default_pg, new_pg) for the test. When initializing sharded tensors, we only initialize (call the constructor) in ranks of the new_pg, but the `_init_rpc` call in `post_init()` will try to all_gather in the default_pg, in this case, rank 0 does not initialize(call constructor), so it does not have this all_gather, thus it hangs. 

https://github.com/pytorch/pytorch/blob/eb2a803406397e884c0d7999f782fa82e0204348/torch/distributed/_sharded_tensor/api.py#L220

I can think of a solution on this but not entirely sure the correct semantic:

we use `sharded_tensor.process_group.all_gather` instead of `rpc.api._all_gather` here, so that it only broadcast the sharded tensor ids in the new_pg, later we still share `all_tensor_ids` in default_pg world.

wanchaol replied:
@pritamdamania87 I thought about this more and the solution i proposed might not work, as for those ranks who don't know about sharded tensor initialization, it does not have the `_register_remote_shards` or sharded tensor map on it at all. In this case, I am thinking that can we only broadcast and register remote shards in the `new_group` instead of the default_group?

https://github.com/pytorch/pytorch/blob/e963b436913f8324a12f8ff0987be7b4085020cb/torch/distributed/_sharded_tensor/api.py#L634

pritamdamania87 replied:
@wanchaol Yes I do feel we should do the all_gather only within the ranks of the `new_group`. Would this work out or are there other issues? I think we can potentially fix this in a follow PR to unblock landing this critical fix for the test suite.

wanchaol replied:
yeah i think this would work out, i will work on a follow up PR after this land.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1904,10 +1886,6 @@ def test_init_from_local_shards_invalid_local_shards(self):
             placement=f"rank:{self.rank}/cuda:{self.rank}"
         )
 
-        empty_local_shards = []
-        with self.assertRaisesRegex(ValueError, 'have no local shards on all ranks'):
-            sharded_tensor = _sharded_tensor.init_from_local_shards(empty_local_shards, [10, 10], init_rrefs=True)
-
```
Is this test no longer relevant?

wanchaol replied:
There were some issues when I try to run this test, mostly because if everywhere is none, we ended up having the `local_shards_device` be CPU, but we are calling `all_gather_object` on NCCL pg, let me see what I can do to make this legit.

wanchaol replied:
just updated with removing the `local_shards_device` and check with `isinstance` instead, I think this should make it cleaner :)

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -440,10 +440,10 @@ def _raise_if_mismatch(expected, actual, prop_name, rank, is_property=False):
             assert shard_meta in local_shard_metadatas, \
                 "local shard metadata not in sharded_tensor_metadata!"
 
+            _raise_if_mismatch(tensor_properties.layout, local_shard_tensor.layout, "layout", current_rank, True)
```
nit: Any reason to move this line?

pritamdamania87 reviewed (commented):
Code changes look good, mostly had questions regarding some of the unit test changes.

Commit:
Update on "[shard] fix with_comms wrapper"


When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
with_comms
def test_case:
    ...
```
or
```
with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[shard] fix with_comms wrapper"


When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
with_comms
def test_case:
    ...
```
or
```
with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[shard] fix with_comms wrapper"


When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
with_comms
def test_case:
    ...
```
or
```
with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

wanchaol commented:
ci down because of sev https://github.com/pytorch/pytorch/issues/69722, not related to the PR
